Add loop parallelization utility to ExecutorsUtils

pcadabam opened the pull request:
We find loop parallelizing logic in several parts of the code base hence trying to make a standard way of doing it.
Added a utility method to parallelize loops. Applies a {@link Function} to every element in a {@link List} in parallel by spawning threads. A list containing the result obtained by applying the function is returned. The method is a blocking call and will wait for all the elements in the list to be processed or timeoutInSecs which ever is earlier.

@chavdar can you review?

chavdar commented on gobblin-utility/src/main/java/gobblin/util/ExecutorsUtils.java:
```diff
@@ -146,4 +154,68 @@ public static void shutdownExecutorService(ExecutorService executorService, Opti
+   *
+   * @throws IllegalArgumentException if input list or function is null
+   */
+  public static <T, V> List<T> parallelize(List<V> list, final Function<V, T> function, int threadCount,
```
1) I think inputOutput is unnecessary. Your task logic can be along the lines of result.set(i, function(list.get(i)) if result is pre-allocated.

2) You need to handle errors in function() or timeouts. You need to use submit() instead of execute() and store all Futures. Once all tasks are done, you need to check if all tasks completed successfully. If any of them failed or timed out, parallelize() needs to through an exception.

pcadabam replied:
Done

chavdar commented on gobblin-utility/src/test/java/gobblin/util/ExecutorsUtilsTest.java:
```diff
@@ -47,4 +52,22 @@ public void run() {
+
+  @Test
+  public void testParallelize() {
+    List<Integer> nums = Lists.newArrayList();
```
I'd add more elements than threads.

pcadabam replied:
Done

chavdar commented on gobblin-utility/src/test/java/gobblin/util/ExecutorsUtilsTest.java:
```diff
@@ -47,4 +52,22 @@ public void run() {
+    List<String> result = ExecutorsUtils.parallelize(nums, multiply , 2, 60, Optional.<Logger>absent());
+    Assert.assertEquals(Arrays.asList("15","50"), result);
+  }
 }
```
We need tests for the following cases:
* A function call throws an Exception
* A function call times out

pcadabam replied:
Done

sahilTakiar commented:
@chavdar this may be a separate discussion, but are we ready to migrate to Java 8? Then we get this for free https://docs.oracle.com/javase/tutorial/collections/streams/parallelism.html#executing_streams_in_parallel

ibuenros commented:
I don't think we should migrate to java 8 yet. As of spring 2015 (latest
stats I could find in a quick search), java 8 adoption was at 20% according
to https://plumbr.eu/blog/java/java-version-statistics-2015-edition
Given that gobblin is open source, I think we should wait until there is
further adoption of java 8 before migrating.

On Fri, Oct 9, 2015 at 12:47 PM, sahilTakiar <notifications@github.com>
wrote:

> @chavdar <https://github.com/chavdar> this may be a separate discussion,
> but are we ready to migrate to Java 8? Then we get this for free
> https://docs.oracle.com/javase/tutorial/collections/streams/parallelism.html#executing_streams_in_parallel
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/linkedin/gobblin/pull/376#issuecomment-146971297>.
>

pcadabam commented:
@chavdar addressed all the comments. Please take a look.
@sahilTakiar Yes later versions of java have similar kind of utils. Some of the code used to wait for future tasks is made easy in java7 http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorCompletionService.html
But I think we will need this until we migrate to java8.

chavdar commented on gobblin-utility/src/main/java/gobblin/util/ExecutorsUtils.java:
```diff
@@ -146,4 +155,73 @@ public static void shutdownExecutorService(ExecutorService executorService, Opti
+    Preconditions.checkArgument(list != null, "Input list can not be null");
+    Preconditions.checkArgument(function != null, "Function can not be null");
+
+    final List<T> results = Lists.newArrayList();
```
nit: This is a small optimization. I'd probably use "new ArrayList<T>(list.size())". Thus we can avoid the dynamic growing of the list as we already now the final size.

chavdar commented:
One small comment. Overall, LGTM.
